src/python/lib/ensembl/compara/filesys/filecmp.py

```diff
@@ -0,0 +1,61 @@
+"""
+Copyright [1999-2015] Wellcome Trust Sanger Institute and the EMBL-European Bioinformatics Institute
+Copyright [2016-2020] EMBL-European Bioinformatics Institute
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import filecmp
+from pathlib import Path
+
+from Bio import Phylo
+
+from .dircmp import PathLike
+
+
+# File extensions that should be interpreted as the same file format:
+NEWICK_EXT = {'.nw', '.nwk', '.newick', '.nh'}
+
+
+def file_cmp(fpath1: PathLike, fpath2: PathLike) -> bool:
+    """Returns True if files `fpath1` and `fpath2` are equivalent, False otherwise.
+
+    Args:
+        fpath1: First file path.
+        fpath2: Second file path.
+
+    """
+    fext1 = Path(fpath1).suffix
+    fext2 = Path(fpath2).suffix
+    if (fext1 in NEWICK_EXT) and (fext2 in NEWICK_EXT):
+        return _tree_cmp(fpath1, fpath2)
+    # Resort to a shallow binary file comparison (files with identical os.stat() signatures are taken to be
+    # equal)
+    return filecmp.cmp(str(fpath1), str(fpath2))
+
+
+def _tree_cmp(fpath1: PathLike, fpath2: PathLike, tree_format: str = 'newick') -> bool:
+    """Returns True if trees stored in `fpath1` and `fpath2` are equivalent, False otherwise.
+
+    Args:
+        fpath1: First tree file path.
+        fpath2: Second tree file path.
+        tree_format: Tree format, i.e. ``newick``, ``nexus``, ``phyloxml`` or ``nexml``.
+
+    """
+    ref_tree = Phylo.read(fpath1, tree_format)
+    target_tree = Phylo.read(fpath2, tree_format)
+    # Both trees are considered equal if they have the same leaves and the same distance from each to the root
+    ref_dists = {leaf.name: ref_tree.distance(leaf) for leaf in ref_tree.get_terminals()}
+    target_dists = {leaf.name: target_tree.distance(leaf) for leaf in target_tree.get_terminals()}
```

Comment:
`ref_dists` and `target_dists` are dictionaries with the format `leaf_name: distance_to_root`. So when comparing these two dictionaries in the next line (`ref_dists == target_dists`), it will only return `True` if the keys are the same (same leaves), and the distance of each leaf to the root is the same.

Reply:
Ah I did miss that. Just to check, does `get_terminals()` return the leaves in any specific (alphabetical?) order? If not, I think the comparison will fail? The trees should be allowed to have rotated ancestral nodes, without affecting the relationships in the phylogeny

Reply:
This should be per leaf, regardless of the topology. In fact, the only problem is that, for instance, one star tree and one tree with internal nodes, both with the same leaves, will be considered equal if the distance to the root for each leaf is the same.

I have left the dictionary comparison as it is supposed to be the fastest way (I assume there will be C code behind the curtains for this).

Reply:
The newick format forces a rooted tree. This means the root, purely for newick formatting, could be placed anywhere, and it would not be incorrect, just different, because we do not expect a rooted tree necessarily. In your comment you suggest that the distance per leaf is calculated from leaf to root. In an unrooted tree, with a root placed elsewhere, whilst the topology is otherwise the same, the distance returned for that leaf can be different. This is why in the first check, a total tree distance was compared, rather than one between root and leaf. Perhaps then, a check should be implemented to first check if the tree is expected to be rooted or not, if it is not expected to be rooted, only the leaf names and the total tree distance should be compared between the two files?

Reply:
(If the leaf distance is not between root and leaf, but between leaf and direct ancestral node, then it is just the comment that needs changing, but a total tree distance additionally still needs to be calculated)

Reply:
After talking with Cristi, we have decided to leave this method as it is for now, as it should work for the species tree. The ideas we have discussed are good for a more general tree/phylogeny comparison method, but it should be added when other pipelines that require a more general tree comparison are included in CITest.